server/server.go

```diff
@@ -269,3 +290,70 @@ func logRotate(lf *os.File) {
 		log.Printf("crypki: rotated logfile: %s", name)
 	}
 }
+
+type shutdownCounterConfig struct {
+	consecutiveCountLimit int32
+	timeRangeCountLimit   int32
+	tickerDuration        time.Duration
+
+	// The function is provided by users to shutdown the server. This function is
+	// guaranteed to run only once.
+	shutdownFn func()
+}
+
+type shutdownCounter struct {
+	config shutdownCounterConfig
+
+	// consecutiveCounter keeps track of the number of consecutive failures with
+	// Internal grpc code.
+	consecutiveCounter int32
+
+	// timeRangeCounter counts the number of failures with Internal grpc code in
+	// the most recent 1 minutes.
+	timeRangeCounter int32
+
+	// once ensures that the shutdown function only runs once.
+	once sync.Once
+}
+
+func (c *shutdownCounter) shutdown() {
+	c.once.Do(func() {
+		if c.config.shutdownFn != nil {
+			c.config.shutdownFn()
+		}
+	})
+}
+
+func (c *shutdownCounter) startTicker() {
+	timer := time.NewTicker(c.config.tickerDuration)
+	for range timer.C {
+		if c.timeRangeCounter >= c.config.timeRangeCountLimit {
+			go c.shutdown()
+			timer.Stop()
+			return
+		}
+		// reset counter
+		atomic.StoreInt32(&c.timeRangeCounter, 0)
+	}
+}
+
+func (c *shutdownCounter) Interceptor(code codes.Code) {
+	if code == codes.Internal {
+		atomic.AddInt32(&c.consecutiveCounter, 1)
+		atomic.AddInt32(&c.timeRangeCounter, 1)
+	} else {
+		// reset counter
+		atomic.StoreInt32(&c.consecutiveCounter, 0)
+	}
+	if c.consecutiveCounter >= c.config.consecutiveCountLimit {
+		go c.shutdown()
+	}
+}
+
+func newShutdownCounter(config shutdownCounterConfig) *shutdownCounter {
+	counter := &shutdownCounter{
+		config: config,
+	}
+	go counter.startTicker()
+	return counter
+}
```

Comment:
Maybe we can create a new package (directory) `interceptor`, create a new file there, and move this part of code to the newly created file? The `interceptor*` files can also be moved there. Otherwise, both `shutdownCounter.Interceptor` and `StatusInterceptor` doesn't need to be exported.